apps/keyboard/js/imes/latin/latin.js

```diff
@@ -1252,22 +1252,19 @@
     return c === '.' || c === '?' || c === '!';
   }
 
-  function selectionChange(detail) {
-    // We would get selectionchange event when the user type each char,
-    // or accept a word suggestion, so don't update suggestions in these
-    // cases.
-    if (detail.ownAction) {
+  function stateChange(state) {
+    var pendingSelection =
+      (state.selectionEnd > state.selectionStart) ? state.selectionEnd : 0;
+
+    if (cursor === state.selectionStart &&
+        selection === pendingSelection &&
+        inputText === state.text) {
       return;
     }
 
-    //XXX: Don't update inputText here, since textBeforeCursor would only
-    // contain 100 chars at most.
-    cursor = detail.selectionStart;
-    if (detail.selectionEnd > detail.selectionStart) {
-      selection = detail.selectionEnd;
-    } else {
-      selection = 0;
-    }
+    cursor = state.selectionStart;
+    selection = pendingSelection;
+    inputText = state.text;
```

Comment:
I think the key point is here.

The behavior of clearing text is different in <input> and contenteditable. We expect inputcontextchange event but get only selectionchnage, so we can not update text accordingly.

This unified handler updated inputText even on selectionchnage, so our prediction function can work with correct inputText and correct cursor.